components/common/infinite_scroll.jsx

```diff
@@ -0,0 +1,142 @@
+         */
+        bufferValue: PropTypes.number,
+
+        /**
+         * The total number of items to be scrolled through
+         */
+        totalItems: PropTypes.number.isRequired,
+
+        /**
+         * The number of items to load in a single fetch
+         */
+        itemsPerPage: PropTypes.number.isRequired,
+
+        /**
+         * The current page that has been scrolled to
+         */
+        pageNumber: PropTypes.number.isRequired,
+    };
+
+    static defaultProps = {
+        bufferValue: SCROLL_BUFFER,
+        endOfDataMessage: '',
+        styleClass: '',
+    };
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            isFetching: false,
+            isEndofData: false,
+        };
+        this.node = React.createRef();
+    }
+
+    componentDidMount() {
+        this.node.current.addEventListener('scroll', this.debounceHandleScroll);
+    }
+
+    componentWillUnmount() {
+        this.node.current.removeEventListener('scroll', this.debounceHandleScroll);
+    }
+
+    validateBuffer = (buffer) => {
+        if (buffer < SCROLL_BUFFER) {
+            return SCROLL_BUFFER;
+        }
+        return Math.abs(buffer);
+    }
+
+    getAmountOfPages = (total, freq) => {
+        return Math.ceil(total / freq);
+    }
+
+    handleScroll = () => {
+        const {isFetching, isEndofData} = this.state;
+        const {callBack, bufferValue, totalItems, itemsPerPage, pageNumber} = this.props;
+
+        const node = this.node.current;
+        const validBuffer = this.validateBuffer(bufferValue);
+
+        const toScroll = node.scrollHeight - node.clientHeight - validBuffer;
+        const nearBottom = node.scrollTop > toScroll;
+
+        if (nearBottom && !isEndofData && !isFetching) {
+            this.setState({isFetching: true},
+                async () => {
+                    await callBack();
+
+                    this.setState({
+                        isFetching: false,
+                    });
+
+                    if (totalItems === 0) {
+                        return;
+                    }
+
+                    const amountOfPages = this.getAmountOfPages(totalItems, itemsPerPage);
+
+                    if (pageNumber === amountOfPages) {
+                        this.setState({
+                            isEndofData: true,
+                        });
+                    }
+                });
+        }
+    }
+
+    debounceHandleScroll = debounce(this.handleScroll, DEBOUNCE_WAIT_TIME);
+
+    render() {
+        const {children, endOfDataMessage, styleClass} = this.props;
+        const {isFetching, isEndofData} = this.state;
+        return (
+            <>
+                <div
+                    className={`infinite-scroll ${styleClass}`}
+                    ref={this.node}
+                >
+                    {children}
+                </div>
+                {isFetching && (
+                    <LoadingSpinner text='Loading'/>
```

Comment:
Sorry for the delayed response on this again.
Regarding spinner:

What we want is to have the spinner as part of the list at the bottom. So this way it looks the same as other infinite lists we already have like center channel, search in RHS etc

```
                <div
                    className={`infinite-scroll ${styleClass}`}
                    ref={this.node}
                >
                    {children}
                    {!isEndofData && <LoadingScreen />}
                </div>
                {isEndofData && endOfDataMessage}
```

Something like the above will work though you might need to pass styles to `LoadingScreen` to constraint the height to be not too big. What we usually do is to have the loader at the bottom if we know there are more results to be loaded that way it does not flicker to show the loader when reaching to the bottom of the list. The loaded is removed when all the results are loaded

If we want to achieve this i think we need to assign the right value to `isEndofData` in constructor so that dont show the loader when there are only 5 teams or so. For this either you can pass in the value from parent or deduce it from `pageNumber` `totalItems` and `itemsPerPage`

Reply:
@sudheerDev currently working to implement this feedback.

Reply:
@sudheerDev I had implemented this earlier, maybe slightly different because of the Loader I was using. Made changes to align more with this review :)

Reply:
Can we have the loader always and not show it only when `isEndofData` is set?. As of now this is based of `isFetching` which will toggle every-time we reach to the bottom of the list. 

In effect it when user scrolls really fast they would reach to the end of the list then they see a loader added to the bottom. Instead if we add it to the list and keep it there always there wont be any toggle of loader. we can remove it when we reach `isEndofData` making it feel like other instances we have and the UX we prefer.

Reply:
> @sudheerDev I had implemented this earlier, maybe slightly different because of the Loader I was using. Made changes to align more with this review :)

I am not sure if you understood my first comment in the review thread here https://github.com/mattermost/mattermost-webapp/pull/3165#discussion_r322649144 so i made a new one to bring it to your attention.

Reply:
@AGMETEOR Hey, I tried this and we still have the loader only when fetching teams.

The way it works is that if user scroll down he can see the end of list and then we make API call and add loader at the bottom.

Let me know if you need me to push a commit on top if i am unclear on this.